src/rgw/rgw_admin.cc

```diff
@@ -1061,6146 +67,1088 @@ class StoreDestructor {
-    return ret;
-  }
-  if (commit) {
-    ret = commit_period(realm, period, remote, url, access, secret, force);
-    if (ret < 0) {
-      cerr << "failed to commit period: " << cpp_strerror(-ret) << std::endl;
-      return ret;
-    }
-  }
-  encode_json("period", period, formatter);
-  formatter->flush(cout);
-  return 0;
-}
-
-static int init_bucket_for_sync(const string& tenant, const string& bucket_name,
-                                const string& bucket_id, rgw_bucket& bucket)
-{
-  RGWBucketInfo bucket_info;
-
-  int ret = init_bucket(tenant, bucket_name, bucket_id, bucket_info, bucket);
-  if (ret < 0) {
-    cerr << "ERROR: could not init bucket: " << cpp_strerror(-ret) << std::endl;
-    return ret;
-  }
-
-  return 0;
-}
+  rgw_user_init(store);
+  rgw_bucket_init(store->meta_mgr);
 
-static int do_period_pull(RGWRESTConn *remote_conn, const string& url,
-                          const string& access_key, const string& secret_key,
-                          const string& realm_id, const string& realm_name,
-                          const string& period_id, const string& period_epoch,
-                          RGWPeriod *period)
-{
-  RGWEnv env;
-  req_info info(g_ceph_context, &env);
-  info.method = "GET";
-  info.request_uri = "/admin/realm/period";
-
-  map<string, string> &params = info.args.get_params();
-  if (!realm_id.empty())
-    params["realm_id"] = realm_id;
-  if (!realm_name.empty())
-    params["realm_name"] = realm_name;
-  if (!period_id.empty())
-    params["period_id"] = period_id;
-  if (!period_epoch.empty())
-    params["epoch"] = period_epoch;
-
-  bufferlist bl;
-  JSONParser p;
-  int ret = send_to_remote_or_url(remote_conn, url, access_key, secret_key,
-                                  info, bl, p);
-  if (ret < 0) {
-    cerr << "request failed: " << cpp_strerror(-ret) << std::endl;
-    return ret;
-  }
-  ret = period->init(g_ceph_context, store, false);
-  if (ret < 0) {
-    cerr << "faile to init period " << cpp_strerror(-ret) << std::endl;
-    return ret;
-  }
-  try {
-    decode_json_obj(*period, &p);
-  } catch (JSONDecoder::err& e) {
-    cout << "failed to decode JSON input: " << e.message << std::endl;
-    return -EINVAL;
-  }
-  ret = period->store_info(false);
-  if (ret < 0) {
-    cerr << "Error storing period " << period->get_id() << ": " << cpp_strerror(ret) << std::endl;
-  }
-  // store latest epoch (ignore errors)
-  period->update_latest_epoch(period->get_epoch());
-  return 0;
-}
+  StoreDestructor store_destructor(store);
 
-static int read_current_period_id(RGWRados* store, const std::string& realm_id,
-                                  const std::string& realm_name,
-                                  std::string* period_id)
-{
-  RGWRealm realm(realm_id, realm_name);
-  int ret = realm.init(g_ceph_context, store);
-  if (ret < 0) {
-    std::cerr << "failed to read realm: " << cpp_strerror(-ret) << std::endl;
+  if (raw_storage_op) {
```

Comment:
here using an abstract class and a factory pattern will help remove this huge switch.
see rgw_op.h : you will need a handler class and an op class (in this case admin command class) for each command type (user, period, bucket , ...)

Reply:
Will look, thanks!